From 4ce1a91fb219f570b0871071a2ec8ac97d90c0bc Mon Sep 17 00:00:00 2001
From: "kingstom.chen" <kingstom.chen@gmail.com>
Date: Thu, 15 May 2025 16:04:46 +0800
Subject: [PATCH] build: add options to override subproject paths

this allows downstream builds to specify custom GLM and stb header paths.
---
 meson.build       | 21 +++++++++++++++++----
 meson_options.txt |  2 ++
 2 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/meson.build b/meson.build
index 802e4a0..2c3d426 100644
--- a/meson.build
+++ b/meson.build
@@ -50,8 +50,23 @@ dep_x11 = dependency('x11')
 dep_wayland = dependency('wayland-client')
 vulkan_dep = dependency('vulkan')
 
-glm_dep = dependency('glm')
-stb_dep = dependency('stb')
+glm_include_dir = get_option('glm_include_dir')
+if glm_include_dir != ''
+  glm_inc = include_directories(glm_include_dir)
+  glm_dep = declare_dependency(include_directories : glm_inc)
+else
+  glm_proj = subproject('glm')
+  glm_dep = glm_proj.get_variable('glm_dep')
+endif
+
+stb_include_dir = get_option('stb_include_dir')
+if stb_include_dir != ''
+  stb_inc = include_directories(stb_include_dir)
+  stb_dep = declare_dependency(include_directories : stb_inc)
+else
+  stb_proj = subproject('stb')
+  stb_dep = stb_proj.get_variable('stb_dep')
+endif
 
 if get_option('enable_openvr_support')
   openvr_dep = dependency('openvr', version: '>= 2.7', required : false)
diff --git a/meson_options.txt b/meson_options.txt
index 27302d65b..acb716b09 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -8,3 +8,5 @@ option('enable_gamescope',           type : 'boolean', value : true, description
 option('enable_gamescope_wsi_layer', type : 'boolean', value : true, description: 'Build Gamescope layer')
 option('enable_openvr_support',      type : 'boolean', value : true, description: 'OpenVR Integrations')
 option('benchmark', type: 'feature', description: 'Benchmark tools')
+option('glm_include_dir',            type : 'string',  value : '', description: 'Custom path to override GLM headers')
+option('stb_include_dir',            type : 'string',  value : '', description: 'Custom path to override STB headers')
\ No newline at end of file
